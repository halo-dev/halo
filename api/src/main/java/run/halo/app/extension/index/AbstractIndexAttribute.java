package run.halo.app.extension.index;

import lombok.EqualsAndHashCode;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;
import run.halo.app.extension.GVK;

@EqualsAndHashCode
public abstract class AbstractIndexAttribute<E extends Extension> implements IndexAttribute {
    private final Class<E> objectType;

    /**
     * Creates a new {@link AbstractIndexAttribute} for the given object type.
     *
     * @param objectType must not be {@literal null}.
     */
    public AbstractIndexAttribute(Class<E> objectType) {
        Assert.notNull(objectType, "Object type must not be null");
        Assert.state(isValidExtension(objectType),
            "Invalid extension type, make sure you have annotated it with @" + GVK.class
                .getSimpleName());
        this.objectType = objectType;
    }

    @Override
    public Class<E> getObjectType() {
        return this.objectType;
    }

    boolean isValidExtension(Class<? extends Extension> type) {
        return type.getAnnotation(GVK.class) != null;
    }
}
