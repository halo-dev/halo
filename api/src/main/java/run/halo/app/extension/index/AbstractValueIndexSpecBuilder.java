package run.halo.app.extension.index;

import org.springframework.util.Assert;
import run.halo.app.extension.Extension;

/**
 * Abstract base implementation of {@link IndexSpecBuilder} for value indexes.
 *
 * @param <E> the type of extension
 * @param <K> the type of key
 * @param <B> the type of builder
 * @author johnniang
 * @since 2.22.0
 */
abstract class AbstractValueIndexSpecBuilder<
    E extends Extension,
    K extends Comparable<K>,
    B extends IndexSpecBuilder<E, K, B>
    > implements IndexSpecBuilder<E, K, B> {

    protected final String name;

    protected final Class<K> keyType;

    protected boolean unique = false;

    protected boolean nullable = true;

    protected AbstractValueIndexSpecBuilder(String name, Class<K> keyType) {
        Assert.hasText(name, "Index name must not be blank");
        Assert.notNull(keyType, "Key type must not be null");
        this.name = name;
        this.keyType = keyType;
    }

    public B unique(boolean unique) {
        this.unique = unique;
        return (B) this;
    }

    public B nullable(boolean nullable) {
        this.nullable = nullable;
        return (B) this;
    }

}
