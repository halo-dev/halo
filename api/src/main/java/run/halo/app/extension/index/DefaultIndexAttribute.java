package run.halo.app.extension.index;

import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;

@Deprecated(forRemoval = true, since = "2.22.0")
class DefaultIndexAttribute<E extends Extension, K extends Comparable<K>>
    implements IndexAttribute<E, K> {

    private final Class<E> objectType;

    private final Class<K> keyType;

    private final Function<E, Set<K>> valuesFunc;

    public DefaultIndexAttribute(Function<E, Set<K>> valuesFunc, Class<E> objectType,
        Class<K> keyType) {
        Assert.notNull(valuesFunc, "Values function must not be null");
        Assert.notNull(objectType, "Cannot resolve object type");
        Assert.notNull(keyType, "Cannot resolve key type");
        this.valuesFunc = valuesFunc;
        this.objectType = objectType;
        this.keyType = keyType;
    }

    @Override
    public Class<E> getObjectType() {
        return objectType;
    }

    @Override
    public Class<K> getKeyType() {
        return keyType;
    }

    @Override
    public Set<K> getValues(E e) {
        if (!checkType(e)) {
            throw new IllegalArgumentException("Object type does not match");
        }
        return Optional.ofNullable(this.valuesFunc.apply(e)).orElse(Set.of());
    }

    private boolean checkType(Extension object) {
        return getObjectType().isInstance(object);
    }
}
