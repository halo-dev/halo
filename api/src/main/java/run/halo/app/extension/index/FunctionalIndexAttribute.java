package run.halo.app.extension.index;

import java.util.Set;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import org.springframework.lang.Nullable;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;

@EqualsAndHashCode(callSuper = true)
public class FunctionalIndexAttribute<E extends Extension>
    extends AbstractIndexAttribute<E> {

    @EqualsAndHashCode.Exclude
    private final Function<E, String> valueFunc;

    /**
     * Creates a new {@link FunctionalIndexAttribute} for the given object type and value function.
     *
     * @param objectType must not be {@literal null}.
     * @param valueFunc value function must not be {@literal null}.
     */
    public FunctionalIndexAttribute(Class<E> objectType,
        Function<E, String> valueFunc) {
        super(objectType);
        Assert.notNull(valueFunc, "Value function must not be null");
        this.valueFunc = valueFunc;
    }

    @Override
    public Set<String> getValues(Extension object) {
        var value = getValue(object);
        return value == null ? Set.of() : Set.of(value);
    }

    /**
     * Gets the value for the given object.
     *
     * @param object the object to get the value for.
     * @return returns the value for the given object.
     */
    @Nullable
    public String getValue(Extension object) {
        if (getObjectType().isInstance(object)) {
            return valueFunc.apply(getObjectType().cast(object));
        }
        throw new IllegalArgumentException("Object type does not match");
    }
}
