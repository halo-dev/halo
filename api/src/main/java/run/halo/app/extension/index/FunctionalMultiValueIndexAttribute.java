package run.halo.app.extension.index;

import java.util.Set;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;
import run.halo.app.extension.Unstructured;

@EqualsAndHashCode(callSuper = true)
public class FunctionalMultiValueIndexAttribute<E extends Extension>
    extends AbstractIndexAttribute<E> {

    @EqualsAndHashCode.Exclude
    private final Function<E, Set<String>> valueFunc;

    /**
     * Creates a new {@link FunctionalIndexAttribute} for the given object type and value function.
     *
     * @param objectType object type must not be {@literal null}.
     * @param valueFunc value function must not be {@literal null}.
     */
    public FunctionalMultiValueIndexAttribute(Class<E> objectType,
        Function<E, Set<String>> valueFunc) {
        super(objectType);
        Assert.notNull(valueFunc, "Value function must not be null");
        this.valueFunc = valueFunc;
    }

    @Override
    public Set<String> getValues(Extension object) {
        if (object instanceof Unstructured unstructured) {
            var ext = Unstructured.OBJECT_MAPPER.convertValue(unstructured, getObjectType());
            return getNonNullValues(ext);
        }
        if (getObjectType().isInstance(object)) {
            return getNonNullValues(getObjectType().cast(object));
        }
        throw new IllegalArgumentException("Object type does not match");
    }

    private Set<String> getNonNullValues(E object) {
        var values = valueFunc.apply(object);
        return values == null ? Set.of() : values;
    }
}
