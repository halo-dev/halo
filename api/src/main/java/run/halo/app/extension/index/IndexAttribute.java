package run.halo.app.extension.index;

import java.util.Set;
import run.halo.app.extension.Extension;

public interface IndexAttribute {

    /**
     * Specify this class is belonged to which extension.
     *
     * @return the extension class.
     */
    Class<? extends Extension> getObjectType();

    /**
     * Get the value of the attribute.
     *
     * @param object the object to get value from.
     * @param <E> the type of the object.
     * @return the value of the attribute must not be null.
     */
    <E extends Extension> Set<String> getValues(E object);
}
