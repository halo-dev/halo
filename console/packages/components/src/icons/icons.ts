import IconDashboard from "~icons/ri/dashboard-3-line";
import IconArrowLeft from "~icons/ri/arrow-left-s-line";
import IconArrowRight from "~icons/ri/arrow-right-s-line";
import IconArrowDown from "~icons/ri/arrow-down-s-fill";
import IconArrowUpLine from "~icons/ri/arrow-up-s-line";
import IconArrowDownLine from "~icons/ri/arrow-down-s-line";
import IconBookRead from "~icons/ri/book-read-line";
import IconPages from "~icons/ri/pages-line";
import IconMessage from "~icons/ri/message-3-line";
import IconPalette from "~icons/ri/palette-line";
import IconListSettings from "~icons/ri/list-settings-line";
import IconMagic from "~icons/ri/magic-line";
import IconUserSettings from "~icons/ri/user-settings-line";
import IconSettings from "~icons/ri/settings-4-line";
import IconPlug from "~icons/ri/plug-2-line";
import IconEye from "~icons/ri/eye-line";
import IconEyeOff from "~icons/ri/eye-off-line";
import IconFolder from "~icons/ri/folder-2-line";
import IconMore from "~icons/ri/more-line";
import IconClose from "~icons/ri/close-line";
import IconErrorWarning from "~icons/ri/error-warning-line";
import IconCheckboxCircle from "~icons/ri/checkbox-circle-line";
import IconInformation from "~icons/ri/information-line";
import IconCloseCircle from "~icons/ri/close-circle-line";
import IconDeleteBin from "~icons/ri/delete-bin-2-line";
import IconAddCircle from "~icons/ri/add-circle-line";
import IconSave from "~icons/ri/save-line";
import IconList from "~icons/ri/list-unordered";
import IconGrid from "~icons/ri/grid-line";
import IconCheckboxFill from "~icons/ri/checkbox-circle-fill";
import IconSearch from "~icons/ri/search-2-line";
import IconComputer from "~icons/ri/computer-line";
import IconPhone from "~icons/ri/smartphone-line";
import IconTablet from "~icons/ri/tablet-line";
import IconUserFollow from "~icons/ri/user-follow-line";
import IconExchange from "~icons/ri/exchange-line";
import IconGitHub from "~icons/ri/github-fill";
import IconUpload from "~icons/ri/upload-cloud-2-line";
import IconShieldUser from "~icons/ri/shield-user-line";
import IconGitBranch from "~icons/ri/git-branch-line";
import IconStopCircle from "~icons/ri/stop-circle-line";
import IconForbidLine from "~icons/ri/forbid-line";
import IconCodeBoxLine from "~icons/ri/code-box-line";
import IconDatabase2Line from "~icons/ri/database-2-line";
import IconTeam from "~icons/ri/team-fill";
import IconCharacterRecognition from "~icons/ri/character-recognition-line";
import IconCalendar from "~icons/ri/calendar-line";
import IconLink from "~icons/ri/link";
import IconUserLine from "~icons/ri/user-line";
import IconMotionLine from "~icons/ri/emotion-line";
import IconReplyLine from "~icons/ri/reply-line";
import IconExternalLinkLine from "~icons/ri/external-link-line";
import IconRefreshLine from "~icons/ri/refresh-line";
import IconWindowLine from "~icons/ri/window-line";
import IconSendPlaneFill from "~icons/ri/send-plane-fill";
import IconRocketLine from "~icons/ri/rocket-line";
import IconArrowUpCircleLine from "~icons/ri/arrow-up-circle-line";
import IconArrowDownCircleLine from "~icons/ri/arrow-down-circle-line";
import IconTerminalBoxLine from "~icons/ri/terminal-box-line";
import IconClipboardLine from "~icons/ri/clipboard-line";
import IconLockPasswordLine from "~icons/ri/lock-password-line";
import IconServerLine from "~icons/ri/server-line";
import IconRiPencilFill from "~icons/ri/pencil-fill";
import IconZoomInLine from "~icons/ri/zoom-in-line";
import IconZoomOutLine from "~icons/ri/zoom-out-line";
import IconArrowLeftRightLine from "~icons/ri/arrow-left-right-line";
import IconArrowUpDownLine from "~icons/ri/arrow-up-down-line";
import IconRiUpload2Fill from "~icons/ri/upload-2-fill";
import IconNotificationBadgeLine from "~icons/ri/notification-badge-line";
import IconLogoutCircleRLine from "~icons/ri/logout-circle-r-line";
import IconAccountCircleLine from "~icons/ri/account-circle-line";
import IconSettings3Line from "~icons/ri/settings-3-line";

export {
  IconDashboard,
  IconArrowLeft,
  IconArrowRight,
  IconArrowDown,
  IconBookRead,
  IconPages,
  IconMessage,
  IconPalette,
  IconListSettings,
  IconMagic,
  IconUserSettings,
  IconSettings,
  IconPlug,
  IconEye,
  IconEyeOff,
  IconFolder,
  IconMore,
  IconClose,
  IconErrorWarning,
  IconCheckboxCircle,
  IconInformation,
  IconCloseCircle,
  IconDeleteBin,
  IconAddCircle,
  IconSave,
  IconList,
  IconGrid,
  IconCheckboxFill,
  IconArrowUpLine,
  IconArrowDownLine,
  IconSearch,
  IconComputer,
  IconPhone,
  IconTablet,
  IconUserFollow,
  IconExchange,
  IconGitHub,
  IconUpload,
  IconShieldUser,
  IconGitBranch,
  IconStopCircle,
  IconForbidLine,
  IconCodeBoxLine,
  IconDatabase2Line,
  IconTeam,
  IconCharacterRecognition,
  IconCalendar,
  IconLink,
  IconUserLine,
  IconMotionLine,
  IconReplyLine,
  IconExternalLinkLine,
  IconRefreshLine,
  IconWindowLine,
  IconSendPlaneFill,
  IconRocketLine,
  IconArrowUpCircleLine,
  IconArrowDownCircleLine,
  IconTerminalBoxLine,
  IconClipboardLine,
  IconLockPasswordLine,
  IconServerLine,
  IconRiPencilFill,
  IconZoomInLine,
  IconZoomOutLine,
  IconArrowLeftRightLine,
  IconArrowUpDownLine,
  IconRiUpload2Fill,
  IconNotificationBadgeLine,
  IconLogoutCircleRLine,
  IconAccountCircleLine,
  IconSettings3Line,
};
