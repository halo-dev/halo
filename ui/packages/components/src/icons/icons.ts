import IconAccountCircleLine from "~icons/ri/account-circle-line";
import IconAddCircle from "~icons/ri/add-circle-line";
import IconArrowDownCircleLine from "~icons/ri/arrow-down-circle-line";
import IconArrowDown from "~icons/ri/arrow-down-s-fill";
import IconArrowDownLine from "~icons/ri/arrow-down-s-line";
import IconArrowLeftRightLine from "~icons/ri/arrow-left-right-line";
import IconArrowLeft from "~icons/ri/arrow-left-s-line";
import IconArrowRight from "~icons/ri/arrow-right-s-line";
import IconArrowUpCircleLine from "~icons/ri/arrow-up-circle-line";
import IconArrowUpDownLine from "~icons/ri/arrow-up-down-line";
import IconArrowUpLine from "~icons/ri/arrow-up-s-line";
import IconBookRead from "~icons/ri/book-read-line";
import IconCalendar from "~icons/ri/calendar-line";
import IconCharacterRecognition from "~icons/ri/character-recognition-line";
import IconCheckboxFill from "~icons/ri/checkbox-circle-fill";
import IconCheckboxCircle from "~icons/ri/checkbox-circle-line";
import IconClipboardLine from "~icons/ri/clipboard-line";
import IconCloseCircle from "~icons/ri/close-circle-line";
import IconClose from "~icons/ri/close-line";
import IconCodeBoxLine from "~icons/ri/code-box-line";
import IconComputer from "~icons/ri/computer-line";
import IconDashboard from "~icons/ri/dashboard-3-line";
import IconDatabase2Line from "~icons/ri/database-2-line";
import IconDeleteBin from "~icons/ri/delete-bin-2-line";
import IconMotionLine from "~icons/ri/emotion-line";
import IconErrorWarning from "~icons/ri/error-warning-line";
import IconExchange from "~icons/ri/exchange-line";
import IconExternalLinkLine from "~icons/ri/external-link-line";
import IconEye from "~icons/ri/eye-line";
import IconEyeOff from "~icons/ri/eye-off-line";
import IconFolder from "~icons/ri/folder-2-line";
import IconForbidLine from "~icons/ri/forbid-line";
import IconGitBranch from "~icons/ri/git-branch-line";
import IconGitHub from "~icons/ri/github-fill";
import IconGrid from "~icons/ri/grid-line";
import IconHistoryLine from "~icons/ri/history-line";
import IconImageAddLine from "~icons/ri/image-add-line";
import IconInformation from "~icons/ri/information-line";
import IconLink from "~icons/ri/link";
import IconListSettings from "~icons/ri/list-settings-line";
import IconList from "~icons/ri/list-unordered";
import IconLockPasswordLine from "~icons/ri/lock-password-line";
import IconLogoutCircleRLine from "~icons/ri/logout-circle-r-line";
import IconMagic from "~icons/ri/magic-line";
import IconMessage from "~icons/ri/message-3-line";
import IconMore from "~icons/ri/more-line";
import IconNotificationBadgeLine from "~icons/ri/notification-badge-line";
import IconPages from "~icons/ri/pages-line";
import IconPalette from "~icons/ri/palette-line";
import IconRiPencilFill from "~icons/ri/pencil-fill";
import IconPlug from "~icons/ri/plug-2-line";
import IconRefreshLine from "~icons/ri/refresh-line";
import IconReplyLine from "~icons/ri/reply-line";
import IconRocketLine from "~icons/ri/rocket-line";
import IconSave from "~icons/ri/save-line";
import IconSearch from "~icons/ri/search-2-line";
import IconSendPlaneFill from "~icons/ri/send-plane-fill";
import IconServerLine from "~icons/ri/server-line";
import IconSettings3Line from "~icons/ri/settings-3-line";
import IconSettings from "~icons/ri/settings-4-line";
import IconShieldUser from "~icons/ri/shield-user-line";
import IconPhone from "~icons/ri/smartphone-line";
import IconStopCircle from "~icons/ri/stop-circle-line";
import IconTablet from "~icons/ri/tablet-line";
import IconTeam from "~icons/ri/team-fill";
import IconTerminalBoxLine from "~icons/ri/terminal-box-line";
import IconTimerLine from "~icons/ri/timer-line";
import IconToolsFill from "~icons/ri/tools-fill";
import IconRiUpload2Fill from "~icons/ri/upload-2-fill";
import IconUpload from "~icons/ri/upload-cloud-2-line";
import IconUserFollow from "~icons/ri/user-follow-line";
import IconUserLine from "~icons/ri/user-line";
import IconUserSettings from "~icons/ri/user-settings-line";
import IconWindowLine from "~icons/ri/window-line";
import IconZoomInLine from "~icons/ri/zoom-in-line";
import IconZoomOutLine from "~icons/ri/zoom-out-line";

export {
  IconAccountCircleLine,
  IconAddCircle,
  IconArrowDown,
  IconArrowDownCircleLine,
  IconArrowDownLine,
  IconArrowLeft,
  IconArrowLeftRightLine,
  IconArrowRight,
  IconArrowUpCircleLine,
  IconArrowUpDownLine,
  IconArrowUpLine,
  IconBookRead,
  IconCalendar,
  IconCharacterRecognition,
  IconCheckboxCircle,
  IconCheckboxFill,
  IconClipboardLine,
  IconClose,
  IconCloseCircle,
  IconCodeBoxLine,
  IconComputer,
  IconDashboard,
  IconDatabase2Line,
  IconDeleteBin,
  IconErrorWarning,
  IconExchange,
  IconExternalLinkLine,
  IconEye,
  IconEyeOff,
  IconFolder,
  IconForbidLine,
  IconGitBranch,
  IconGitHub,
  IconGrid,
  IconHistoryLine,
  IconImageAddLine,
  IconInformation,
  IconLink,
  IconList,
  IconListSettings,
  IconLockPasswordLine,
  IconLogoutCircleRLine,
  IconMagic,
  IconMessage,
  IconMore,
  IconMotionLine,
  IconNotificationBadgeLine,
  IconPages,
  IconPalette,
  IconPhone,
  IconPlug,
  IconRefreshLine,
  IconReplyLine,
  IconRiPencilFill,
  IconRiUpload2Fill,
  IconRocketLine,
  IconSave,
  IconSearch,
  IconSendPlaneFill,
  IconServerLine,
  IconSettings,
  IconSettings3Line,
  IconShieldUser,
  IconStopCircle,
  IconTablet,
  IconTeam,
  IconTerminalBoxLine,
  IconTimerLine,
  IconToolsFill,
  IconUpload,
  IconUserFollow,
  IconUserLine,
  IconUserSettings,
  IconWindowLine,
  IconZoomInLine,
  IconZoomOutLine,
};
