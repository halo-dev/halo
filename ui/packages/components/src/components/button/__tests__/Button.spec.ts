import { describe, expect, it, vi } from "vitest";
import { VButton } from "../index";
import { mount } from "@vue/test-utils";
import { IconSettings } from "../../../icons/icons";

describe("Button", () => {
  it("should render", () => {
    expect(mount(VButton).html()).contains("button");
    expect(mount(VButton).html()).toMatchSnapshot();
  });

  it("should render with text", () => {
    expect(mount(VButton, { slots: { default: "Hello Halo" } }).text()).toBe(
      "Hello Halo"
    );
  });

  it("should work with type prop", () => {
    // default button type is default
    expect(mount(VButton).find(".btn").classes()).toContain("btn-default");

    ["primary", "secondary", "danger"].forEach((type) => {
      const button = mount(VButton, { props: { type } });

      expect(button.html()).toMatchSnapshot();
      expect(button.find(".btn").classes()).toContain(`btn-${type}`);
      button.unmount();
    });
  });

  it("should work with size prop", async () => {
    // default button size is md
    expect(mount(VButton).find(".btn").classes()).toContain("btn-md");

    ["lg", "sm", "xs"].forEach((size) => {
      const button = mount(VButton, { props: { size } });

      expect(button.html()).toMatchSnapshot();
      expect(button.find(".btn").classes()).toContain(`btn-${size}`);
      button.unmount();
    });
  });

  it("should work with circle prop", async () => {
    const button = mount(VButton);

    // default: false
    expect(button.find(".btn").classes()).not.toContain("btn-circle");

    await button.setProps({ circle: true });
    expect(button.html()).toMatchSnapshot();
    expect(button.find(".btn").classes()).toContain("btn-circle");
  });

  it("should work with block prop", async () => {
    const button = mount(VButton);

    // default: false
    expect(button.find(".btn").classes()).not.toContain("btn-block");

    await button.setProps({ block: true });
    expect(button.html()).toMatchSnapshot();
    expect(button.find(".btn").classes()).toContain("btn-block");
  });

  it("should work with disabled prop", async () => {
    const onClick = vi.fn(() => 1);

    // default: false
    const button = mount(VButton, {
      emits: { click: onClick },
    });

    await button.trigger("click");
    expect(onClick).toHaveBeenCalled();

    onClick.mockReset();
    await button.setProps({ disabled: true });
    await button.trigger("click");

    expect(button.html()).toMatchSnapshot();
    expect(onClick).not.toHaveBeenCalled();
  });

  it("should work with loading prop", async function () {
    const wrapper = mount({
      data() {
        return {
          loading: true,
        };
      },
      template: `
        <v-button :loading="loading" >
          Hello
        </v-button>
      `,
      components: {
        VButton,
      },
    });

    expect(wrapper.find(".btn").classes()).toContain("btn-loading");
    expect(wrapper.find(".btn-icon").exists()).toBe(true);

    // set loading = false
    await wrapper.setData({ loading: false });
    expect(wrapper.find(".btn").classes()).not.toContain("btn-loading");
    expect(wrapper.find(".btn-icon").exists()).toBe(false);
  });

  it("should work with loading prop and icon slot", async function () {
    const wrapper = mount({
      data() {
        return {
          loading: false,
        };
      },
      template: `
        <v-button :loading="loading">
        <template #icon>
          IconSettings
        </template>
        Hello
        </v-button>
      `,
      components: {
        VButton,
        IconSettings,
      },
    });

    expect(wrapper.find(".btn-icon").exists()).toBe(true);
    expect(wrapper.find(".btn").classes()).not.toContain("btn-loading");

    await wrapper.setData({ loading: true });

    expect(wrapper.find(".btn-icon").exists()).toBe(true);
    expect(wrapper.find(".btn").classes()).toContain("btn-loading");
  });
});
