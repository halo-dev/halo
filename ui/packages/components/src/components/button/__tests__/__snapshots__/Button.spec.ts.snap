// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should render 1`] = `
"<button class=\\"btn-md btn-default btn\\" type=\\"button\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;

exports[`Button > should work with block prop 1`] = `
"<button class=\\"btn-md btn-default btn-block btn\\" type=\\"button\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;

exports[`Button > should work with circle prop 1`] = `
"<button class=\\"btn-md btn-default btn-circle btn\\" type=\\"button\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;

exports[`Button > should work with disabled prop 1`] = `
"<button class=\\"btn-md btn-default btn\\" type=\\"button\\" disabled=\\"\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;

exports[`Button > should work with size prop 1`] = `
"<button class=\\"btn-lg btn-default btn\\" type=\\"button\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;

exports[`Button > should work with size prop 2`] = `
"<button class=\\"btn-sm btn-default btn\\" type=\\"button\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;

exports[`Button > should work with size prop 3`] = `
"<button class=\\"btn-xs btn-default btn\\" type=\\"button\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;

exports[`Button > should work with type prop 1`] = `
"<button class=\\"btn-md btn-primary btn\\" type=\\"button\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;

exports[`Button > should work with type prop 2`] = `
"<button class=\\"btn-md btn-secondary btn\\" type=\\"button\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;

exports[`Button > should work with type prop 3`] = `
"<button class=\\"btn-md btn-danger btn\\" type=\\"button\\">
  <!--v-if--><span class=\\"btn-content\\"></span>
</button>"
`;
